### PR TITLE
[BD-21] Fix requirements installation

### PR DESCRIPTION
**Pull request by regisb**
Running `pip install -r requirements/dev.txt` was failing because `-e .`
was expanded to an absolute path. This is a long-standing issue in
pip-tools: https://github.com/jazzband/pip-tools/issues/204

So we got rid of the "-e ." line in dev.in and recompiled the
requirement files. We realized that this did not pull all requirements
because some of them were hardcoded in setup.cfg. We decided to move
those requirements to a base.in file and to dynamically load this file
from setup.py. This removes the need to manually upgrade a few
dependencies. To achieve that, we had to add a MANIFEST.in file.

Then, in order to fix tests we had to include test requirements, so we
created a test.in requirement file.

Note that the `py35-pylint24` still fails because black==20.8b1 is
incompatible with Python 3.5. AFAIK there is no version compatible with
Python 3.5.

cc @robrap @cpennington

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @regisb! I've created [BLENDED-633](https://openedx.atlassian.net/browse/BLENDED-633) to keep track of it in Jira.   More details are on the [BD-21](https://openedx.atlassian.net/wiki/spaces/COMM/pages/1596358943/BD-21+Toggles+Settings+Documentation) project page.


When this pull request is ready, tag your edX technical lead.

File: requirements/dev.txt
Comment (regisb) [inline requirements/dev.txt:7]: This is the line that prompted this PR cc @cpennington

Comment (robrap) [reply requirements/dev.txt:7]: Should we assume this was just a mistake and remove it?

Comment (regisb) [reply requirements/dev.txt:7]: The problem is that if we add `-e .` to the `dev.in` file, it will automatically re-compile to an absolute path in the `dev.txt` file because of https://github.com/jazzband/pip-tools/issues/204.
Note that the cookiecutter does not add `-e .` to dev.in (for this good reason): https://github.com/edx/edx-cookiecutters/blob/master/cookiecutter-django-app/%7B%7Bcookiecutter.repo_name%7D%7D/requirements/dev.in

Comment (robrap) [reply requirements/dev.txt:7]: I am assuming that merging both `-e .` and `-e file:///home/cpennington/work/code/edx-lint` were just mistakes.  Right?

File: setup.py
Comment (robrap) [inline setup.py:54]: I'm pretty lazy when it comes to reviewing. I didn't realize you were just copying this code from cookiecutter.  Is that correct, or do I need to carefully review this?

File: requirements/base.in
Comment (robrap) [inline requirements/base.in:6]: No line break at end of file.

Comment (robrap) [inline requirements/base.in:6]: Any idea what constraints are needed at all, and which are needed here vs in a constraints.txt? See https://github.com/edx/edx-cookiecutters/blob/3079f93e1f72a7d5339e2340fd4a13e49af2c4eb/cookiecutter-django-app/%7B%7Bcookiecutter.repo_name%7D%7D/requirements/base.in#L2
Note: minimum constraints are fine here.  It is the pinning and maximum constraints that I am wondering about.

File: requirements/test.in
Comment (robrap) [inline requirements/test.in:5]: Missing line break at end of file.

File: requirements/base.in
Comment (robrap) [inline requirements/base.in:2]: Feels like `Django` would be a base requirement if `pylint-django` is. Not sure how far you want to go with this PR, but we usually let tox manage the Django versions for testing.
See:
* https://github.com/edx/edx-cookiecutters/blob/3079f93e1f72a7d5339e2340fd4a13e49af2c4eb/cookiecutter-django-app/%7B%7Bcookiecutter.repo_name%7D%7D/Makefile#L46-L48
* Also, search for "django" in https://github.com/edx/edx-cookiecutters/blob/3079f93e1f72a7d5339e2340fd4a13e49af2c4eb/cookiecutter-django-app/%7B%7Bcookiecutter.repo_name%7D%7D/tox.ini

File: Makefile
Comment (robrap) [inline Makefile:32]: @nedbat: Any idea why `base.in` wasn't added when this first started using pip-compile?

File: requirements/dev.txt
Comment (regisb) [reply requirements/dev.txt:7]: I guess that adding `-e .` was intended, but the compilation to an absolute path wasn't.

File: setup.py
Comment (regisb) [reply setup.py:54]: This comes from [edx-toggles' setup.py](https://github.com/edx/edx-toggles/blob/master/setup.py) but I'll copy this from the [cookiecutter](https://github.com/edx/edx-cookiecutters/blob/master/python-template/%7B%7Bcookiecutter.placeholder_repo_name%7D%7D/setup.py) instead.

File: requirements/base.in
Comment (regisb) [reply requirements/base.in:6]: I tried to be backward compatible, so I just copy-pasted the content from setup.cfg. In the future, requirements should not be pinned in the base.in requirement file.

To be honest, I don't understand the point of having a contraints.txt when we can simply have a base.in. I guess that constraints.txt are for temporary contraints?

File: requirements/base.in
Comment (regisb) [reply requirements/base.in:2]: Again, I tried to be backward-compatible here, so I did not introduce any new requirement.

File: Makefile
Comment (regisb) [reply Makefile:32]: Here is the original conversation: https://github.com/edx/edx-lint/pull/103#discussion_r329625807

**Comment by regisb** [comment]
Related to this change: CI fails because it cannot install black on python 3.5. @feanil I see that you introduced black as an edx-lint requirement here: https://github.com/edx/edx-lint/pull/108
Black isn't used anywhere in this repo; in particular, the quality jobs don't run black. Does it mean that it's safe to remove it from the requirements?

File: requirements/base.in
Comment (robrap) [reply requirements/base.in]: Yes. The pins and upper constraints can all get properly documented for when and how they can be removed.

Maybe just add a TODO comment to clarify the next step?

Comment (regisb) [reply requirements/base.in]: :ok_hand:

File: requirements/base.in
Comment (robrap) [reply requirements/base.in:3]: That makes sense. I tend to want to leave notes to my future self around things like this, but I’ll leave that up to you.

File: Makefile
Comment (robrap) [reply Makefile:32]: Still not quite following what has changed between now and then that this couldn’t have been the original solution?

Also, Cale had a comment around pinning pylint that probably should move to a comment next to the pin.

File: requirements/dev.txt
Comment (cpennington) [reply requirements/dev.txt:7]: Oops. Yeah, definitely didn't mean to add an absolute path here.

File: Makefile
Comment (cpennington) [reply Makefile:32]: Gah. Hoisted by my own petard here, it looks like. I totally forgot the original considerations around this. That said, we should definitely do something so that we can use a standard `make upgrade` flow.

**Comment by feanil** [comment]
@regisb I'm fine with dropping black if it's causing issues though I might suggest dropping Python 3.5 support instead since that version of python has gone out of support.  This repo was an early test to see if there would be any major pushback to black formatting and to see what it looked like in a real edx repo.

**Comment by cpennington** [comment]
I'm surprised that we'd need to drop black to get it installable on py3.5. Since black is only used in `dev.in`/`dev.txt`, I'd hope that you'd only need black when you were developing edx-lint. Also, it seems like having it running in tox inside a py3.8 environment as part of the quality build would be a step forward.

File: requirements/base.in
Comment (robrap) [inline requirements/base.in:1]: I know you only added this TODO for me, but now I am going to ask you to remove it. :)
In the other PR you mentioned, @cpennington had written:
> since edx-lint is intended to standardize pylint versions, that setup.py should pin particular pylint versions.
Rather than blocking this PR on the right wording for this comment, let's just remove the comment for now.

**Review by robrap — APPROVED** [review]
@regisb: After you remove the mentioned comment, you can squash how you want it, and I can merge.

@feanil: If you want to add back black and fix for Python 3.8, feel free to handle that it in a separate PR. Thanks!

File: MANIFEST.in
Comment (robrap) [inline MANIFEST.in:1]: @regisb: I missed this.  Please add new line here as well.  Thanks.

**Comment by regisb** [comment]
> UPDATE: You could also do any of the following:

@robrap this PR has already grown beyond what I thought it would become, so I'd rather leave the responsibility to others to implement those changes (or implement those in separate PRs).

> Update CHANGELOG.rst. It is out of date. See https://github.com/edx/edx-lint/releases

I have no idea what was released as part of v1.5.1 and v1.5.2, so I'd rather avoid adding words that might prove controversial. 

> Update version if you think it is worthwhile. There should be no net changes, but it might be worthwhile ensuring it all builds correctly.

I don't want to make a no-op release just for testing purposes.

> Maybe add a pull request template to remind people to update CHANGELOG.rst?

This project does not include a PR template yet, so adding one here is very much outside the scope of this PR.

**Comment by robrap** [comment]
@regisb: I think you are implicitly saying that you don't think this change requires a note in the changelog, even for unreleased. I'm fine with that if that is what you mean.

Also, I think this is ready to merge from your end, but it helps when you note that explicitly, so I know you aren't planning on squashing or have any other timing issues around me merging your code.  Please confirm and I will merge. Thanks.

FYI: I've updated my general changelog ideas here: https://github.com/edx/edx-lint/pull/129

**Comment by regisb** [comment]
> I think you are implicitly saying that you don't think this change requires a note in the changelog, even for unreleased. I'm fine with that if that is what you mean.
> Also, I think this is ready to merge from your end, but it helps when you note that explicitly, so I know you aren't planning on squashing or have any other timing issues around me merging your code. Please confirm and I will merge. Thanks.

Yes, this is correct. Sorry about the confusion!

**Pull request merged**
